[PR] Overall Typescript refactor

LuchoTurtle opened the pull request:
Seeing that @thomaux already did an awesome job with implementing type definitions, I wanted to learn more about generics in Typescript so I'm **attempting** to rewrite the main `index.js` file in Typescript so it's friendlier for everyone 😄 . 

This is very much a work in progress, as I'm trying to use the pre-existing type definitions in the current package. There's much to be done but I guess this is a start. I'm running every modification against tests so functions that are typed so far should remain functional.

If anyone would like to chime in and give some feedback, that would be lovely!

nelsonic commented:
@LuchoTurtle thanks very much for opening this PR as we discussed. 👌 
100% on-board with a **`TypeScript`** re-write of this module for greater `TS` support. 🎉 
(_obvs it will still be compiled to JS for full backward compatibility ..._)
Please consider opening an issue to outline the work that needs to be done. 🆕 
Preferably with a checklist so that it can be worked on over multiple sessions. 👩‍💻 
Thanks! 🙏

LuchoTurtle commented:
@thomaux thank you, sir! I was guided by your awesome type definitions. I try to reuse as much as possible but I gotta admit I'm stumped on some aspects, especially when I constantly get ``M` type cannot be used as an index type`. 

Again, I'm new to this level of TS generics, so I'm just trying to learn. But thank you for the feedback, I'll take that into account 😄

nelsonic commented:
Relevant reading: https://blog.liblab.com/typescript-npm-packages-done-right/ 👀

nelsonic commented:
@LuchoTurtle remind me why this was paused? ⏸️ 💭 ⏳

LuchoTurtle commented:
@nelsonic I was only meant to work on this whenever I "had the time" or "didn't have any outstanding issues to work at", which really hasn't happened since.
Additionally, I'm not particularly well-versed in TS generic types and I keep stumbling upon blocking errors that are a bit beyond me. I'm thinking of doing something like https://www.totaltypescript.com/ to actually sort this PR out but it's a big task and there are other more important tasks to tackle first 👌

nelsonic commented:
@LuchoTurtle Thanks for pairing on this during the morning today. 🧑‍💻 👌 
Even if it felt like "slow progress", don't worry. 💭 
It's a worthy effort if you want to add `TypeScript` skills 💪 
& maintaining/refactoring a highly popular library to your list. ✅

LuchoTurtle commented:
I'm having trouble creating a type that yields all the combinations of the `client` and their properties, e.g `DynamoDB.DocumentClient`. 
I've tried several approaches but I don't seem to get it to work.
Created a question on https://stackoverflow.com/questions/77413867/how-to-create-a-type-that-yields-a-dot-notation-of-nested-class-properties-as-st with a [playground](https://www.typescriptlang.org/play?#code/PTAEBEE8DsEMFsD24BCBYAUAYwDawM76gD64iWArvAKbQAuAwjgJa12gDeAvpnDfgAdYWaiTKUa9Jq3qceGTABNquWACdRqwmJgJkKOUpV4NoPtUHDRpXUlSdMoUNQAeAxGvbN47zxHJUbNJsoAC8YgGSjCxsANyY8pggoACCDACyRqqmWkTEaemGGMrZouaWIiQFDhhOrr7sdJACogzUnswAZsxYsHTUKIiKkGGg+HRqzNAA5vEYiRjJKQDqAMq8CBZClSurNXVuHl4+R6kZo-kZcwcNoN63UHz6F496qNegS2ugABYEoLBQHRENBQIhOmC6D92qBcDI6PgAHTIhKYNEKRZgdIjSjjRDwIHNCxAv7sACSoAA7rBZMDQNNqOx8EdqIpQAA3Tlc7k83no+qnJotUDBegAOU2owA1tRIODCS15btYp8wFDmERKR4pfh+Yc-ELRGKLP1FKK6BKaAAeBjOFz9aCKIjmy3UAB80tl8sNSrWAG0GABdFXJfA-RAUHCKaAAckaPw1oEgrCjoAAROJAlIYvQ06AAD7ptodbq9fqDYZpgD86KSYAp1NpiFAACNRLAWzhRHSGexAYaqcwoQDQBRoMwQWCIeNJjNQCx+mpYDgiCwZaAAAZpgqI4t0Lo9PoDIaQPOFjO2fSIzNRc1pzACg1E0DG8as80AMUjOFdNrtDqdEUcwtTYABoX1tVwAKIV9TRdTYbTdD1wg3AASDgGC4RF0LFTCNxVWsMSAA) for anyone to try and give it a whirl.

LuchoTurtle commented:
Okay, so I've made several changes and I believe this is a good starting base to migrate this package to `Typescript`.
I don't really know *why* the coverage is down 10% in here. I'm running `npm test` on my computer and it's yielding 99%.

![image](https://github.com/dwyl/aws-sdk-mock/assets/17494745/ab1175a1-82c6-4f84-b344-f311e5281618)

This is in itself a bit weird since I'm using code that is very similar, so the code flow should not have changed (or at least I think I made the effort to do so).

Nonetheless, here's a list of changes that have been implemented:

- The `tsconfig` file uses Matt Pocock's recommended settings (https://www.totaltypescript.com/tsconfig-cheat-sheet), which enforce strictness and correct transpiling to `js` files according to the target platform.
- Made some formatting changes to the README so it's easier for people to differentiate between using the package for `AWS SDK for Javascript v2` and `v3`.
- Added `src` folder and put all the code there. Now there's only `index.ts` (which stems from the previous `index.js`) and `types.js`, which holds all the types that are being used throughout the package that can also be exported.
- Made changes to the `package.json` scripts, having added a `build` script that builds the transpiled `js` files from `ts` files.
- Added `jsdoc` to all the functions inside `index.ts`.
- Made my best effort to type out all the test cases inside `test/index.test.ts`. Some are unfortunately using the `any` because I can't really control user input and cannot infer types from `sinons` stubs/proxies, which highly depends on the function that the user inputs when mocking.
- I've tried using the types that already existed to the max. However, some did not work, so some were refactored and/or not used entirely (the latter being the case of `NestedClassNames` and `NestedMethodNames`). I've added comments to the code so this is better discernable. 
- Typed all of the functions in `index.ts`.
- I've tried maintaining all of the original code flow. However, it was not possible since Typescript enforces stricter undefined behaviour handling.


I believe some things can be improved, though. I wager that we can add a specific PR to better automate publishing and also updating dependencies. This can be done through [`changesets`](https://github.com/un-ts/changesets-gitlab). I'd be happy to do this once this PR is merged and I get approval for it.

nelsonic commented:
@LuchoTurtle this is looking good. 👍 
Let's take a look at it together on our morning call. 👀

LuchoTurtle commented:
So the PR passes on an instance that runs `Node 16.x` but not on `Node 20.x`.

https://github.com/dwyl/aws-sdk-mock/actions/runs/6806283033/job/18507294821?pr=277

Which is weird. I'm running `Node 20.x` and it works on my `localhost`.

![image](https://github.com/dwyl/aws-sdk-mock/assets/17494745/1303da62-1538-4134-a9fc-ab4b541bc51b)

LuchoTurtle commented:
I don't know why the pipeline is failing on this PR. 

I'm using the **exact same tests** (the only difference is that they're typed) and they actually cover more lines % and branch % than the current `main` branch.

Running `npm test` on both branches:

`main`:

![image](https://github.com/dwyl/aws-sdk-mock/assets/17494745/af40af60-8e7a-4ade-9d50-8c3d4e17bd45)

`typescript` branch:

![image](https://github.com/dwyl/aws-sdk-mock/assets/17494745/3fbe2f8e-4af3-4839-b2e9-6102dc4acb6d)

@nelsonic do you have any idea why this could be happening?

LuchoTurtle commented:
Yep, this is ridiculous. I can't figure out why the process on Github Actions is ending with code 1. The tests are all passing, similarly to other actions on `.js`, yet the command ends successfully.

I've tried checking for errors from what I've seen in https://stackoverflow.com/questions/70970086/tests-pass-on-github-actions-but-process-exit-code-1 but I can't find any that would make `tape` or `nyc` execute the `npm test` command with `code` as `1`.

There's absolutely nothing to go by. I've tried lowering the coverage threshold, as from the past two commits, but those seem to have absolutely no effect. 

I'm inclined on just trashing `nyc`/`tap` and use another testing framework, because there's no logs or errors to go by.

LuchoTurtle commented:
I'm going to try https://stackoverflow.com/questions/50459872/no-coverage-nyc-mocha tomorrow 👍 

> There is an issue with the nyc (Instanbul) loader for ESM modules. The recommended solution is to abandon nyc and use the native coverage now built into Node 13 and above: c8.

nelsonic commented:
@LuchoTurtle happy to drop `nyc` and use the native coverage. 👌 
(Will need to do this for other `JS` projects ...)

LuchoTurtle commented:
Because some of the tests are failing on more recent Node versions, it seems it's because we need to use `"type": "module"` in our `package.json`. This will effectively make this library **ES Module-based** (instead of using CommonJS). This change affects how the tests are made.

`v18` of `tap` doesn't seem to want to cooperate with `nyc`. We're stuck with this `tap`'s latest version because it's the only one that works with ES Modules. However, even if we want to ditch `nyc` and only use `tap`, we can't because in `v18`, `tap` [forces users to have either full coverage or not](https://github.com/tapjs/tapjs/issues/964).  

This project has close to 100%. Here's me running `npm run test` on the `main` branch.

![Screenshot 2024-02-27 at 18 46 56](https://github.com/dwyl/aws-sdk-mock/assets/17494745/65447956-e71e-4efc-b31a-d6b4b17f7ac5)

As you can see, there are uncovered lines (outlined in red).

On this branch, I actually got the numbers to be better. Even if the `branches` number is lower - this is because there's a lack of some `else` statements (I tried my best to keep the code intact, so it seems that `nyc` is stricter here). 

![image](https://github.com/dwyl/aws-sdk-mock/assets/17494745/d87132e0-f412-4164-9ece-87b239748bcc)

As you can see, the statements and lines are at 100% (which are arguably the most important parameters). You can run `npm run coverage` to see an HTML representation of the lines that are covereed and *why* the branches are at 83%.

However, `tap` will end with code `1` (indicating an error). `nyc` perceives this as `0%`, and errors out.

It's easier to have a testing framework that allows us to customize thresholds. I also want to keep the original tests (as I did) and focus the PR on migrating Typescript and not adding more code. I feel it's a better compromise to change a few dependencies and use another testing framework.

# What I did

Because of all of these nuances, I'm dropping `tap` and hoping to use [`mocha`](https://github.com/mochajs/mocha) and [`c8`](https://github.com/bcoe/c8) (JS's native engine test engine). I'm using [`expectations`](https://www.npmjs.com/package/expectations) to keep test assertions similar to `jest`'s - similar to https://github.com/traceypooh/js-es-modules-coverage-testing.

All the code now should be `tsd` error-free and the tests should all pass on different environments. I've deleted unnecessary packages.

To build the package, I'm using `tsup`.

LuchoTurtle commented:
I'm now going to update the README with relevant info and submit the PR for review.

LuchoTurtle commented:
@nelsonic this should now be reviewable. To summarize:

- I tried maintaining the original code as close to the original as possible. Of course, this wasn't always possible due to Typescript type-checking.
- there are a few improvements that could have been made with the typing. I'm still learning how this works, so I think I've made fair progress where people can get a minimum of type inference when writing their code. I think this is enough as a "first TS version".
- the tests now use `mocha` and `expectations` instead of `jest`. `nyc` is still used for code coverage. The reason I made this switch was that `nyc` wasn't properly working with `jest` with `ESModule` (which I had to switch from `CommonJS` transpiring in order to pass the tests on different `Node` versions).
- removed unneeded dev dependencies to lean the library out.
- you can see in `package.json` that we are using` type: module`. As it was stated before, this was needed for the package to work on later versions of Node.js. This entailed changing the package to be fully ESModules-based instead of CommonJS. I've tried maintaining the original code to this effect.
- the tests suffered changes because of the change of testing framework. However, the content of the tests are the same (except adding some types so there aren't type errors in them). You can see compare both the old `js` version and this new `ts` one and you'll see that the tests are precisely the same (and they all pass :) ). 
- `src/index.ts` has the original code (with Typescript). `src/types.ts` has types used throughout the `index.ts` file.
- `tsconfig.json` is based from https://www.totaltypescript.com/tsconfig-cheat-sheet. This is from Pocock's, quite literally a Typescript wizard, so these are great settings for the library, I believe.

nelsonic commented on README.md:
```diff
@@ -1,28 +1,34 @@
+
 This module was created to help test AWS Lambda functions but can be used in any situation where the AWS SDK needs to be mocked.
 
+This library is best suited for `AWS SDK for Javascript (v2)` - see the [introductory post on the AWS blog](https://aws.amazon.com/blogs/developer/mocking-modular-aws-sdk-for-javascript-v3-in-unit-tests/) for more context. 
```
We _definitely_ don't want to recommend using a different package ... 😕 
This alternative package only exists because it's taken us too long to update ours. ⌛ 🙃

nelsonic reviewed (dismissed):
@LuchoTurtle updates look good to me. 🙌 
Requested review from @thomaux if he has time. 🤞 
Meanwhile what _version_ should this be published as?
Should it be a _major_ version e.g. `v6`? 💭

LuchoTurtle commented:
@nelsonic 

> Should it be a _major_ version e.g. `v6`? 💭

Without a doubt. Because of the combination of the dependencies being updated and how the code is being transpiled from TS to JS, support only works for Node `v18` upwards.

nelsonic commented:
Ok, cool. Do you want to publish the new version? 📦

LuchoTurtle commented:
I feel like you should be the one to do it, I don't want to mess anything up since it's a package that is highly used :p

nelsonic commented:
![](https://i.giphy.com/wi8Ez1mwRcKGI.webp)

LuchoTurtle commented:
I'm not sure how I'd go about it. Wouldn't this need to be merged first for me to create a new version of the package in `npm`?

nelsonic commented:
Nah, you can create & publish a new version on a branch. 📦 🚀 

Simply run:

```sh
npm version major -m 'Typescript Refactor. see: https://github.com/dwyl/aws-sdk-mock/pull/277"
npm publish
git push && git push --tags
```

Invited you to maintain:
<img width="532" alt="image" src="https://github.com/dwyl/aws-sdk-mock/assets/194400/8ad65477-d733-4cdf-913c-abbf98ece3f4">

Will merge as soon as you publish. :shipit: 
If anything goes "wrong" (highly doubt it) message me on Signal. 💬

LuchoTurtle commented:
I'm intent on publishing this later. I'm at the airport right now and I've fast-forwarded this branch to the `main`. I'm publishing it now because I'm hesitant about having some issues. If that's the case, I could only get back to this in 24 hours.

nelsonic commented:
Do it. I'm here and check GitHub regularly for messages. Not many people will "upgrade" that quickly.

LuchoTurtle commented:
Okay, doing it now (after updating `package-lock.json`.

LuchoTurtle commented:
@nelsonic https://www.npmjs.com/package/aws-sdk-mock **version `6.0.0` published**.

nelsonic reviewed (approved):
Great work @LuchoTurtle 🎉 
Thanks for publishing the new version to NPM. 📦 🚀 
Merging. :shipit:
